File: Difficulty: Medium/0 - 1 Knapsack Problem/0-1-knapsack-problem.py
#User function Template for python3
class Solution:
    
    #Function to return max value that can be put in knapsack of capacity W.
    def get_ans(self, w, idx, wt, val, n, memo):
        if idx < 0:
            return 0
        if memo[idx][w] != -1:
            return memo[idx][w]
        
        a = 0
        if w - wt[idx] >= 0:
            a = val[idx] + self.get_ans(w - wt[idx], idx - 1, wt, val, n, memo)
        
        a = max(a, self.get_ans(w, idx - 1, wt, val, n, memo))
        memo[idx][w] = a
        
        return memo[idx][w]
        
    def knapSack(self,W, wt, val):
        n = len(wt)
        memo = [[-1] * (W + 1) for _ in range(n)]
        return self.get_ans(W, n - 1, wt, val, n - 1, memo)


#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        # n = int(input())
        W = int(input())
        val = list(map(int, input().strip().split()))
        wt = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.knapSack(W, wt, val))

# } Driver Code Ends

File: Difficulty: Medium/0 - 1 Knapsack Problem/0-1-knapsack-problem.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class gfg {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            int w = Integer.parseInt(br.readLine());

            String line = br.readLine();
            String[] tokens = line.split(" ");
            List<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] val = new int[array.size()];
            int idx = 0;
            for (int i : array) val[idx++] = i;

            String lin = br.readLine();
            String[] toke = lin.split(" ");
            List<Integer> array1 = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : toke) {
                array1.add(Integer.parseInt(token));
            }

            int[] wt = new int[array1.size()];
            idx = 0;
            for (int i : array1) wt[idx++] = i;

            // calling method knapSack() of class Solution
            System.out.println(new Solution().knapSack(w, wt, val));
        }
    }
}
// } Driver Code Ends


class Solution {
    
    static int fun(int idx, int wt[], int val[], int w)
    {
        if(w == 0) return 0;
        
        if(idx == 0)
        {
            if(wt[idx] <= w) return val[idx];
            else return 0;
        }
        
        int leftCall = fun(idx - 1, wt, val, w);
        int rightCall = 0;
        
        if(wt[idx] <= w)
        {
            rightCall = val[idx] + fun(idx - 1, wt, val, w - wt[idx]);
        }
        
        return Math.max(leftCall, rightCall);
    }
    // Function to return max value that can be put in knapsack of capacity W.
    static int knapSack(int W, int wt[], int val[]) {
        
        return fun(wt.length-1, wt, val, W);
    }
}
